agreement_dynamic/models/agreement_section.py

```diff
@@ -0,0 +1,254 @@
+    def onchange_copyvalue(self):
+        self.sub_object_id = False
+        self.copyvalue = False
+        if self.field_id and not self.field_id.relation:
+            self.copyvalue = "${{object.{} or {}}}".format(
+                self.field_id.name, self._get_proper_default_value()
+            )
+            self.sub_model_object_field_id = False
+        if self.field_id and self.field_id.relation:
+            self.sub_object_id = self.env["ir.model"].search(
+                [("model", "=", self.field_id.relation)]
+            )[0]
+        if self.sub_model_object_field_id:
+            self.copyvalue = "${{object.{}.{} or {}}}".format(
+                self.field_id.name,
+                self.sub_model_object_field_id.name,
+                self._get_proper_default_value(),
+            )
+
+    @api.onchange("condition_python")
+    def _compute_condition(self):
+        """Compute condition and preview"""
+        for this in self:
+            this.condition_python_preview = this.show_in_report = False
+            if not this.resource_ref_model_id:
+                continue
+            if not this.res_id:
+                continue
+            if not this.condition_python and this.condition_domain == "[]":
+                this.show_in_report = True
+                continue
+            # Use condition_domain here
+            condition_domain = safe_eval(this.condition_domain)
+            record = self.env[this.resource_ref_model_id.model].search(
+                condition_domain, limit=1
+            )
+            if not record:
+                continue
+            if not this.condition_python:
+                this.show_in_report = True
+                continue
+            try:
+                # Check if there are any syntax errors etc
+                this.condition_python_preview = safe_eval(
+                    this.condition_python.strip(), {"object": record}
+                )
+            except Exception as e:
+                # and show debug info
+                this.condition_python_preview = str(e)
+                continue
+            is_valid = bool(this.condition_python_preview)
+            this.show_in_report = is_valid
+            if not is_valid:
+                # Preview is there, but condition is false
+                this.condition_python_preview = "False"
+
+    def _get_proper_default_value(self):
+        self.ensure_one()
+        is_num = self.field_id.ttype in ("integer", "float")
+        value = 0 if is_num else "''"
+        if self.default_value:
+            if is_num:
+                value = "{}"
+            else:
+                value = "'{}'"
+            value = value.format(self.default_value)
+        return value
+
+    # compute the dynamic content for jinja expression
+    def _compute_dynamic_content(self):
+        # a parent with two children
+        h = Header(child=Header(child=Header()))
```

Comment:
Maybe you can split this off into `_get_header_object()` so that it can easily be overrided by child modules.